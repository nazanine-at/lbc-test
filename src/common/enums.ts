export enum EMessageType {
  PUBLIC = 'PUBLIC',
  PRIVATE = 'PRIVATE',
}
