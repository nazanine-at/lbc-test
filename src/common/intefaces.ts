import { EMessageType } from './enums'

export interface IState {
  messages: IMessage[]
  selectedMessage: IMessage
}

export interface IMessage {
  userId?: string
  subject?: string
  message?: string
  viewType?: EMessageType
}
