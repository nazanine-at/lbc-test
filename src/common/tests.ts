import configureMockStore, { MockStoreCreator, MockStoreEnhanced } from 'redux-mock-store'
import { IState } from './intefaces'

const initStore: MockStoreCreator = configureMockStore()
export function mockStore(initialState: IState): MockStoreEnhanced {
  return initStore(initialState)
}
