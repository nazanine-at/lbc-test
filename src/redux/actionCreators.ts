import { IMessage } from '../common/intefaces'

// export const setMessagesCreator = (messages: IMessage[]) => {
//   return { type: 'messages/setMessages', payload: messages }
// }
//
// export const addMessageCreator = (message: IMessage) => {
//   return {
//     type: 'message/addMessage',
//     payload: message,
//   }
// }
//
// export const selectMessageCreator = (message: IMessage | undefined) => {
//   return { type: 'message/selectMessage', payload: message }
// }
//
// export const unselectMessageCreator = () => {
//   return { type: 'message/unselectMessage', payload: {} }
// }
