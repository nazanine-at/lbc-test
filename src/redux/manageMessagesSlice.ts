/**
 * This is a reducer - a function that takes a current state value and an
 * action object describing "what happened", and returns a new state value.
 * A reducer's function signature is: (state, action) => newState
 *
 * The Redux state should contain only plain JS objects, arrays, and primitives.
 * The root state value is usually an object.  It's important that you should
 * not mutate the state object, but return a new object if the state changes.
 *
 * You can use any conditional logic you want in a reducer. In this example,
 * we use a switch statement, but it's not required.
 */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { IState } from '../common/intefaces'

export const fetchMessages = createAsyncThunk('fetchMessages', async () => {
  const res = await fetch('/messages')
  return res.json()
})

export const initialState: IState = {
  messages: [],
  selectedMessage: {},
}

const manageMessagesSlice = createSlice({
  name: 'manageMessages',
  initialState,
  reducers: {
    addMessage: (state: IState, { payload }): void => {
      state.messages = [...state.messages, payload]
    },
    selectMessage: (state: IState, { payload }): void => {
      state.selectedMessage = payload
    },
    unselectMessage: (state: IState): void => {
      state.selectedMessage = {}
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchMessages.fulfilled, (state, { payload }) => {
      state.messages = payload
    })
    builder.addCase(fetchMessages.rejected, (state) => {
      state.messages = []
    })
  },
})

export const { unselectMessage, selectMessage, addMessage } = manageMessagesSlice.actions
export default manageMessagesSlice.reducer
