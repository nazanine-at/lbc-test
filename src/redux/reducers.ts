/**
 * This is a reducer - a function that takes a current state value and an
 * action object describing "what happened", and returns a new state value.
 * A reducer's function signature is: (state, action) => newState
 *
 * The Redux state should contain only plain JS objects, arrays, and primitives.
 * The root state value is usually an object.  It's important that you should
 * not mutate the state object, but return a new object if the state changes.
 *
 * You can use any conditional logic you want in a reducer. In this example,
 * we use a switch statement, but it's not required.
 */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { IState } from '../common/intefaces'

export const fetchMessages = createAsyncThunk('fetchMessages', async () => {
  const res = await fetch('/messages')
  return res.json()
})

const manageMessagesSlice = createSlice({
  name: 'manageMessages',
  initialState: {
    messages: [],
    selectedMessage: {},
  },
  reducers: {
    // @ts-ignore
    setMessages: (state: IState, { payload }): void => {
      state.messages = payload.messages
    },
    // @ts-ignore
    addMessage: (state: IState, { payload }): void => {
      state.messages = [...state.messages, payload]
    },
    // @ts-ignore
    selectMessage: (state: IState, { payload }): void => {
      state.selectedMessage = payload
    },
    // @ts-ignore
    unselectMessage: (state: IState): void => {
      state.selectedMessage = {}
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchMessages.fulfilled, (state, { payload }) => {
      state.messages = payload
    })
    builder.addCase(fetchMessages.rejected, (state, action) => {
      state.messages = []
    })
  },
})

export const { actions, reducer } = manageMessagesSlice
