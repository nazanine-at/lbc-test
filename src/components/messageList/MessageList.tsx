import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import './messageList.css'
import { IMessage, IState } from '../../common/intefaces'
import { EMessageType } from '../../common/enums'
import { selectMessage as selectMessageAction } from '../../redux/manageMessagesSlice'
function MessageList() {
  const messages = useSelector((state: IState) => state.messages)
  const dispatch = useDispatch()
  const [showPrivateMessages, setShowPrivateMessages] = useState(false)

  const selectMessage = (userId: string): void => {
    // find clicked message by userId which should be unique
    const selectedMessage = messages.find(
      (currentMessage: IMessage) => currentMessage.userId === userId
    )
    dispatch(selectMessageAction(selectedMessage))
  }

  const setRestriction = () => {
    // change privacy rule to show or hide private message
    setShowPrivateMessages(!showPrivateMessages)
  }

  if (!messages.length) {
    return null
  }

  return (
    <div className="messageList">
      <button onClick={setRestriction}>
        {showPrivateMessages ? 'Hide Private Messages' : 'Show Private Messages'}
      </button>
      {messages.map(({ subject, message, userId = '', viewType }: IMessage) => {
        if (viewType?.toUpperCase() === EMessageType.PRIVATE && !showPrivateMessages) {
          return null
        }

        return (
          <div key={userId} className="messagePart">
            <button onClick={() => selectMessage(userId)}>Focus the input</button>
            <p>subject: {subject}</p>
            <p>message: {message}</p>
            <p>user: {userId}</p>
            <p>viewType: {viewType}</p>
          </div>
        )
      })}
    </div>
  )
}

export default MessageList
