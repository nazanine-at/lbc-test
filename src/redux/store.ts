import { configureStore } from '@reduxjs/toolkit'
import manageMessagesSlice from './manageMessagesSlice'

export const store = configureStore({ reducer: manageMessagesSlice })
