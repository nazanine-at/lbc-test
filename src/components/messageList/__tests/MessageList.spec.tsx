import React from 'react'
import { render, screen } from '@testing-library/react'
import MessageList from '../MessageList'
import { Provider } from 'react-redux'
import { mockStore } from '../../../common/tests'
import { initialState } from '../../../redux/manageMessagesSlice'
const store = mockStore(initialState)

const { container } = render(
  <Provider store={store}>
    <MessageList />
  </Provider>
)

test('<MessageList/>', () => {
  expect(container).toMatchSnapshot()
})
