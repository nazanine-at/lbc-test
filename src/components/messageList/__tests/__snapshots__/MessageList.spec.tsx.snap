// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageList/> 1`] = `<div />`;
